### PR TITLE
Add emojis:generate_borders Rake task

### PR DESCRIPTION
**Pull request by leo60228**
cc @ThibG 

See #13769, #13766

### CODE REVIEW

**Comment by leo60228** [comment]
Oops, didn't mean to modify the lockfile...

File: lib/tasks/emojis.rake [hunk 1]
```diff
@@ -1,5 +1,45 @@
...
+    end
+    style += " stroke-width: calc(#{old_width} + 4px)"
+    border_elem["style"] = style.strip
+
+    g.add_child(border_elem)
+  end
+  svg.prepend_child(g)
+  File.write(dest, doc.to_xml)
+  puts "Wrote bordered #{codepoint}.svg to #{dest}!"
+end
```
Comment (ThibG) [inline lib/tasks/emojis.rake:41]: CodeClimate complains about a bunch of styling issues.

File: public/emoji/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+*_border.svg
```
Comment (ThibG) [inline public/emoji/.gitignore:1]: My idea was to have the generated files in the repo, like we may do with the lock files, or the `emoji_map.json`.

File: lib/tasks/emojis.rake [hunk 3]
```diff
@@ -55,4 +95,11 @@ namespace :emojis do
     File.write(dest, Oj.dump(map))
     puts "Wrote emojo to destination! (#{dest})"
   end
+
+  desc 'Generate emoji variants with white borders'
+  task :generate_borders, [] do |task, args|
+    for cc in args.extras
+      gen_border cc
+    end
+  end
```
Comment (ThibG) [inline lib/tasks/emojis.rake:104]: My idea was that the unicode characters to generate files from would be hardcoded in here.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
I have made a few requests for changes, but I can take over if you prefer, as I have a good idea what I want to do with the front-end part in particular.

**Comment by leo60228** [comment]
Done.

File: lib/tasks/emojis.rake [hunk 4]
```diff
@@ -55,4 +89,12 @@ namespace :emojis do
     File.write(dest, Oj.dump(map))
     puts "Wrote emojo to destination! (#{dest})"
   end
+
+  desc 'Generate emoji variants with white borders'
+  task :generate_borders do
+    codepoints = %w(1f3b1 1f41c 1f519 26ab 1f5a4 2b1b 25fe 25fc 2712 25aa 1f4a3 1f3b3 1f464 1f465 1f4f7 1f4f8 2663 27b0 1f4b1 1f576 2734 1f50c 1f51a 1f482-200d-2640-fe0f 1f482-1f3fb-200d-2640-fe0f 1f482-1f3fc-200d-2640-fe0f 1f482-1f3fd-200d-2640-fe0f 1f482-1f3fe-200d-2640-fe0f 1f482-1f3ff-200d-2640-fe0f 1f4fd 1f373 1f98d 1f482 1f482-1f3fb 1f482-1f3fc 1f482-1f3fd 1f482-1f3fe 1f482-1f3ff 2714 2797 1f4b2 2796 2716 2795 1f52a 1f573 1f579 1f54b 1f58a 1f58b 1f482-200d-2642-fe0f 1f482-1f3fb-200d-2642-fe0f 1f482-1f3fc-200d-2642-fe0f 1f482-1f3fd-200d-2642-fe0f 1f482-1f3fe-200d-2642-fe0f 1f482-1f3ff-200d-2642-fe0f 1f3a4 1f393 1f3a5 1f3bc 1f51b 1f51c 2660 1f5e3 1f577 1f4de 2122 1f51d 1f3a9 1f983 1f4fc 1f4f9 1f3ae 1f403 1f3f4 3030)
+    codepoints.each do |cc|
+      gen_border cc
+    end
```
Comment (ThibG) [inline lib/tasks/emojis.rake:98]: oh, I was going to suggest something like

```ruby
src = Rails.root.join('app', 'javascript', 'mastodon', 'features', 'emoji', 'emoji_map.json')
    emojis = '🎱🐜⚫🖤⬛◼️◾◼️✒️▪️💣🎳📷📸♣️🕶️✴️🔌💂‍♀️📽️🍳🦍💂🔪🕳️🕹️🕋🖊️🖋️💂‍♂️🎤🎓🎥🎼♠️🎩🦃📼📹🎮🐃🏴👽⚾🐔☁️💨🕊️👀🍥👻🐐❕❔⛸️🌩️🔊🔇📃🌧️🐏🍚🍙🐓🐑💀☠️🌨️🔉🔈💬💭🏐🏳️⚪⬜◽◻️▫️'

    map = Oj.load(File.read(src))

    emojis.each_grapheme_cluster do |emoji|
      gen_border map[emoji]
    end
```

**Comment by leo60228** [comment]
This approach causes artifacting with sharp internal corners. I'm going to see if this approach can be combined with the filter approach.

**Comment by ThibG** [comment]
Thanks again for your work, applying this on top of #13772 and adding a few front-end changes, I get the following for the dark theme:
image

And this for the light theme:
image

It looks pretty much the same on Firefox 76, Chromium 81 and Midori 7, so that is really good! On the very old Firefox 47, the borders do not look as good, but they're still here and make the whole thing legible.

**Comment by leo60228** [comment]
I had very bad artifacting with emoji like the chicken. This tweaked filter is able to hide artifacting without affecting much else, and while seamlessly degrading on browsers without filter support.

**Comment by leo60228** [comment]
The camera with flash emoji looks somewhat odd as one of the corners, despite being internal, is still included in the outline. I'm going to try and tweak it by hand, it shouldn't affect much anyway.

**Comment by ThibG** [comment]
Oh I hadn't noticed you went away from filters to do it purely with the stroke size. Are there browsers without filter support? If not, is there any downside in using only the filter, and not changing the stroke width?

**Comment by leo60228** [comment]
IE9 is the main one. I'm not sure if Mastodon supports IE9. There's no real downside, but it'll require some more tweaking.

**Comment by ThibG** [comment]
Ok. As you wish. Working with stroke width might also benefit from using `stroke-linejoin`

**Comment by leo60228** [comment]
There's tons of small artifacts when using only the filter:
image
Not sure how noticable they'll be zoomed out, though.

**Comment by ThibG** [comment]
Yeah, only using the filter isn't that good. I wonder if `stroke-linejoin: round` doesn't solve all the issues you had which required the filter.
Also I think I get why the borders weren't working as well on Firefox 47: it doesn't handle calc(). Maybe you could use “4px” in all cases, as it seems Twemoji simply doesn't use strokes at all, and thus it will always be 0px.

**Comment by leo60228** [comment]
Thank you for the `stroke-linejoin` tip! It works perfectly.

**Comment by leo60228** [comment]
I got the *opposite* error when running `rubocop` locally...

**Review by ClearlyClaire — APPROVED** [review]

**Comment by leo60228** [comment]
Are you making a separate PR for frontend changes or should I keep this as a draft?

File: lib/tasks/emojis.rake [hunk 5]
```diff
@@ -1,5 +1,36 @@
...
+    border_elem.delete('fill')
+
+    border_elem['stroke'] = 'white'
+    border_elem['stroke-linejoin'] = 'round'
+    border_elem['stroke-width'] = '4px'
+
+    g.add_child(border_elem)
+  end
+  svg.prepend_child(g)
+  svg['style'] = 'background: black;'
```
Comment (ThibG) [inline lib/tasks/emojis.rake:29]: Oops, please remove that :)

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

**Comment by ThibG** [comment]
I think I'll make the front-end changes in another PR. You can mark this PR as ready for review.

**Comment by ThibG** [comment]
Please refresh the generated SVGs as well

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**